### DISCUSSION TITLE
Notifications sent despite form submission isValid set to false

### DISCUSSION THREAD

**Discussion by jrrdnx** [solspace/craft-freeform#1496] [Questions & Answers] [answered]
### What happened?

This issue does involve some custom event hooks, so still unsure if this is actually an issue with Freeform. It is on an older version, but may affect later versions as well.

We have a plugin tying into the `SubmissionsService::EVENT_BEFORE_SUBMIT` event which first checks if the form submission is valid and then reaches out to an API to sync data there first. If that API call should fail, it adds a custom error message and sets `isValid` to false:

```
$form->addError('Failure to submit, testing...');
$event->isValid = false;
```

This code has seemingly been working fine for close to a year, but recently we received a ticket where a user received an email notification from this form but there was no Submission in the Craft control panel. Looking at our custom logs for this API integration, we can see that the API call failed so the form submission would have been marked invalid and the user shown an error. So it appears as if email notifications to Email fields within the form are still triggered even if the form submission is invalid and never completed.

In setting up a simplified test case, I created a form with only an Email field and set up a hook to add an error and set isValid to false:

```
Event::on(
    SubmissionsService::class,
    SubmissionsService::EVENT_BEFORE_SUBMIT,
    function (SubmitEvent $event) {
        $form = $event->getForm();
        $submission = $event->getSubmission();

        if ($form->isValid()) {
            if($form->getHandle() == "test") {
                $form->addError('Failure to submit, testing...');
                $event->isValid = false;
            }
        }

        return;
    }
)
```

### How can we reproduce this?

1. Create a form with an Email field that sends a separate notification from the form's Admin Notification
2. Add a Before Submit event hook to add a form error and set isValid to false
3. Submissions are correctly not saved, but email notifications are still sent.

Additionally, when the Success Behavior is set to Use Return URL the user is directed to this URL and shown the success message when they should instead be shown an error. When the Enable AJAX option is checked, they are correctly shown the error message (screenshot attached) but email notifications are still being sent.

![Screenshot 2024-08-27 at 14 01 46](https://github.com/user-attachments/assets/0758e59c-281d-4a1b-8be8-0ae86537fdac)

### Freeform Edition

Pro

### Freeform Version

3.13.29

### Craft Version

3.8.16

### When did this issue start?

Unsure

**Comment by kjmartens** [comment]
Hi @jrrdnx,

I'll have a developer check into this and see if he has any suggestions or if a fix is necessary. 🙂

**Comment by gustavs-gutmanis** [comment, marked as answer]
Hi @jrrdnx,

I would suggest using the [`Form::EVENT_BEFORE_VALIDATE`](https://docs.solspace.com/craft/freeform/v3/developer/form-events/#on-validating-the-form) event, and perform your API call there, and just set the form error if it is invalid. The rest will be picked up by Freeform, the form won't advance, and an error will be shown.

**Comment by jrrdnx** [comment]
Thanks @gustavs-gutmanis 

We'd still want to use Freeform to validate first though. If there's missing or invalid data in the form, I want that to be fixed first so we're not submitting to the API multiple times with already invalid data.

Would EVENT_AFTER_VALIDATE work better in this case, and if so how would that look different?

**Comment by gustavs-gutmanis** [comment]
Yes @jrrdnx, you can use that event, but then you do need to set `$event->isValid` to `false`.

**Comment by jrrdnx** [comment]
How would I get access to the submission data (which is sent to the API) from the `Form::EVENT_BEFORE_VALIDATE` event?

**Reply by gustavs-gutmanis** [reply]
You would iterate over all `Form` fields like this:

```php
$form = $event->getForm();
foreach ($form->getLayout()->getFields() as $field) {
    $value = $field->getValue();
}
```
